### PR TITLE
[99] Add provider creation task

### PR DESCRIPTION
**Pull request by gpeng**
### Context

We need to onboard users and providers to the Publish/TTAPI sandbox environment.

### Changes proposed in this pull request

Add task that:

* Accepts a CSV
* Creates an organisation
* Creates a provider
* Adds a location to the provider

This task is procedural stylee for now. We are currently discussing building some UI on the API for this sort of support work. This will serve as 'documentation' for creating a better structured service class in the future.

The data set on the provider and associated entities is minimal placeholder data. Sandbox users should probably be directed to update things like the address and location etc otherwise there will be a lot of courses at the same location in Find.

### Guidance to review

The tasks will run locally. Create appropriate CSV files as per the task description.

`bundle exec rake sandbox:create_providers\[./providers.csv\]`

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

**Comment by duncanjbrown** [comment]
Hey @gpeng! This makes sense to me. By way of learning a bit about TTAPI I did some refactoring here — wdyt? Could follow up with the same treatment for the Users import.

**Comment by duncanjbrown** [comment]
For realistic data, should we create a handful of locations along with the provider?

**Comment by gpeng** [comment]
[Sandbox provider rake task](https://trello.com/c/XpRZMAeh/1062-sandbox-provider-rake-task)

**Comment by gpeng** [comment]
> Hey @gpeng! This makes sense to me. By way of learning a bit about TTAPI I did some refactoring here — wdyt? Could follow up with the same treatment for the Users import.

Thanks @duncanjbrown Looks fine. I was holding off splitting it into services for now. If we're going to add a UI for onboarding all users/providers we'll need a non-fake version of the service and we can probably use that for both types. We can extract that later though so happy to go with this approach if you want to start doing that now. The tests are failing on the `env` check though.

**Comment by gpeng** [comment]
> For realistic data, should we create a handful of locations along with the provider?

I was trying to create the minimum possible that would all the user to be able to log in and edit the provider. We need at least one as the UI doesn't handle there being none very well. Another issue with this is the address. If we add them all with the same address then sandbox find is going to end up with a load of courses all at whatever address we choose. Requiring the sandbox users to add them manually might avoid this a bit 😬 I was going to use Faker but that would cause issues in this regard too. Would be good to work out a solution of this.

In the meantime though, if you feel it would be more useful for sandbox users to have a few pre-filled then yes, let's add them.

**Comment by duncanjbrown** [comment]
> > Hey @gpeng! This makes sense to me. By way of learning a bit about TTAPI I did some refactoring here — wdyt? Could follow up with the same treatment for the Users import.
> 
> Thanks @duncanjbrown Looks fine. I was holding off splitting it into services for now. If we're going to add a UI for onboarding all users/providers we'll need a non-fake version of the service and we can probably use that for both types. We can extract that later though so happy to go with this approach if you want to start doing that now. The tests are failing on the `env` check though.

Awesome, thank you — it was a useful learning experience in any case! I'll fix the specs.

**Comment by duncanjbrown** [comment]
> > For realistic data, should we create a handful of locations along with the provider?
> 
> I was trying to create the minimum possible that would all the user to be able to log in and edit the provider. We need at least one as the UI doesn't handle there being none very well. Another issue with this is the address. If we add them all with the same address then sandbox find is going to end up with a load of courses all at whatever address we choose. Requiring the sandbox users to add them manually might avoid this a bit 😬 I was going to use Faker but that would cause issues in this regard too. Would be good to work out a solution of this.
> 
> In the meantime though, if you feel it would be more useful for sandbox users to have a few pre-filled then yes, let's add them.

Nothing to hold this up — let's merge as-is and get feedback?

**Review by defong — CHANGES_REQUESTED** [review]
image

One stone two tech leads... rejected

**Comment by duncanjbrown** [comment]
My first @defong rejection 😊

File: spec/services/providers/create_fake_provider_service_spec.rb [hunk 1]
```diff
@@ -0,0 +1,69 @@
+require "rails_helper"
+
+RSpec.describe Providers::CreateFakeProviderService do
+  let(:provider_name) { "Fake Provider" }
+  let(:provider_code) { "ABC" }
+  let(:provider_type) { "scitt" }
+  let(:recruitment_cycle) { RecruitmentCycle.current.presence || create(:recruitment_cycle) }
```
Comment (defong) [inline spec/services/providers/create_fake_provider_service_spec.rb:7]:
```suggestion
  let(:recruitment_cycle) { find_or_create :recruitment_cycle }
```

Maybe?

File: app/services/providers/create_fake_provider_service.rb [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        address2: provider.address2,
+        address3: provider.address3,
+        address4: provider.address4,
+        postcode: provider.postcode,
+        region_code: provider.region_code,
+      )
+
+      if provider.invalid?
+        errors << "Unable to create Provider #{provider_name}: #{provider.errors.to_sentence}."
+      end
```
[the remark this answers is not in the document]
Comment (duncanjbrown) [reply app/services/providers/create_fake_provider_service.rb:51]: Good point

File: app/services/providers/create_fake_provider_service.rb [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+      errors.empty?
+    end
+
+  private
+
+    def provider_already_exists?
+      if @recruitment_cycle.providers.exists?(provider_name: @provider_name) || \
+          @recruitment_cycle.providers.exists?(provider_code: @provider_code)
```
[the remark this answers is not in the document]
Comment (duncanjbrown) [reply app/services/providers/create_fake_provider_service.rb:60]: Point proven — updating!

File: app/services/providers/create_fake_provider_service.rb [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        address1: provider.address1,
+        address2: provider.address2,
+        address3: provider.address3,
+        address4: provider.address4,
+        postcode: provider.postcode,
+        region_code: provider.region_code,
+      )
+
+      if provider.invalid?
+        errors << "Unable to create Provider #{provider_name}: #{provider.errors.to_sentence}."
```
[the remark this answers is not in the document]
Comment (duncanjbrown) [reply app/services/providers/create_fake_provider_service.rb:50]: It's not actually possible to make an invalid provider with this interface, even if you knock out the name and the code, so I've removed this check

File: app/services/providers/create_fake_provider_service.rb [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    }.freeze
+
+    def initialize(recruitment_cycle:, provider_name:, provider_code:, provider_type:, is_accredited_body:)
+      raise "Can only be run in sandbox or development" unless Rails.env.sandbox? || Rails.env.development? || Rails.env.test?
+
+      @recruitment_cycle = recruitment_cycle
+      @provider_name = provider_name
+      @provider_code = provider_code
+      @provider_type = provider_type
+      @is_accredited_body = is_accredited_body
```
[the remark this answers is not in the document]
Comment (duncanjbrown) [reply app/services/providers/create_fake_provider_service.rb:20]: Added

**Comment by duncanjbrown** [comment]
Should be ready for another look!

File: app/services/providers/create_fake_provider_service.rb [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        location_name: "Site 1",
+        address1: provider.address1,
+        address2: provider.address2,
+        address3: provider.address3,
+        address4: provider.address4,
+        postcode: provider.postcode,
+        region_code: provider.region_code,
+      )
+
+      errors.empty?
```
Comment (defong) [inline app/services/providers/create_fake_provider_service.rb:50]: Does this actually get used?

given 
```ruby
      return false if provider_already_exists?
      return false if attempting_to_make_lead_school_accredited_body?
```

File: spec/services/csv_imports/fake_providers_import_spec.rb [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+      import = described_class.new("csv_file.csv")
+      import.execute
+
+      expect(import.results).to eq(["Created provider Provider A."])
+    end
+  end
+
+  context "two providers to import" do
+    let(:csv_content) do
+      "name,code,type,accredited_body\n\"Provider A\",ABC,scitt,false\n\"Provider B\",DEF,lead_school,true"
```
Comment (defong) [inline spec/services/csv_imports/fake_providers_import_spec.rb:36]:
```suggestion
      "name,code,type,accredited_body\n\"Provider A\",ABC,scitt,false\n\"Provider B\",DEF,lead_school,false"
```

File: app/services/providers/create_fake_provider_service.rb [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+  private
+
+    def provider_already_exists?
+      if @recruitment_cycle.providers.exists?(provider_code: @provider_code)
+        errors << "Provider #{@provider_name} (#{@provider_code}) already exists."
+        true
+      else
+        false
+      end
```
Comment (defong) [inline app/services/providers/create_fake_provider_service.rb:61]:
```suggestion
      end
```

**Review by defong — APPROVED** [review]

File: app/services/providers/create_fake_provider_service.rb (hunk 8 above)
Comment (duncanjbrown) [reply app/services/providers/create_fake_provider_service.rb:61]: I think I prefer the v explicit version?

Comment (defong) [reply app/services/providers/create_fake_provider_service.rb:61]: https://github.com/DFE-Digital/teacher-training-api/blob/be0bba87a456bedc02f6ad21a835ef57133689de/app/services/providers/create_fake_provider_service.rb#L64-L68

it justslooks odd with this underneath

**Pull request merged**